server/broadcast.go

```diff
@@ -460,6 +496,86 @@ func shouldStopSession(err error) bool {
 	return sessionErrRegex.MatchString(err.Error())
 }
 
+func verify(verifier *verification.SegmentVerifier, cxn *rtmpConnection,
+	sess *BroadcastSession, source *stream.HLSSegment,
+	res *net.TranscodeData, URIs []string) error {
+
+	// Cache segment contents if necessary.
+	// If we need to retry transcoding because verification fails,
+	// the the segments' OS location will be overwritten.
+	// Cache the segments so we can restore them in OS if necessary.
+	renditionData := make([][]byte, len(URIs))
+	for i, fname := range URIs {
+		if sess.BroadcasterOS.IsExternal() && drivers.IsOwnExternal(fname) {
+			// If broadcaster is using external storage and segments are there
+			// Then have the verifier use that external storage.
+			continue
+		}
+
+		// Sanity check for absolute URIs to ensure we're actually local.
+		uri, err := url.ParseRequestURI(fname)
+		if err != nil {
+			return err // Implies this is retryable?
+		}
+		memOS, ok := sess.BroadcasterOS.(*drivers.MemorySession)
+		if !uri.IsAbs() && ok {
+			data := memOS.GetData(fname)
+			if data == nil {
+				return errors.New("Missing Local Data")
+			}
+			renditionData[i] = data
+		} else {
+			return errors.New("Expected local storage but did not have it")
+		}
+	}
+
+	params := &verification.Params{
+		ManifestID:   sess.ManifestID,
+		Source:       source,
+		Profiles:     sess.Profiles,
+		Orchestrator: sess.OrchestratorInfo,
+		Results:      res,
+		URIs:         URIs,
+		Renditions:   renditionData,
+	}
+
+	// The return value from the verifier, if any, are the *accepted* params.
+	// The accepted params are not necessarily the same as `params` sent here.
+	// The accepted params may be from an earlier iteration if max retries hit.
+	accepted, err := verifier.Verify(params)
+	if _, retry := err.(verification.Retryable); retry {
+		// If retryable, means tampering was detected from this O
+		// Remove the O from the working set for now
+		// Error falls through towards end if necessary
+		cxn.sessManager.removeSession(sess)
+	}
+	if accepted != nil {
+		// The returned set of results has been accepted by the verifier
+
+		// Check if an earlier verification attempt was the one accepted.
+		// If so, reset the local OS if we're using that since it's been
+		// overwritten with this rendition.
+		if accepted != params && !sess.BroadcasterOS.IsExternal() {
+			for i, data := range accepted.Renditions {
+				// Sanity check that we actually have the rendition data?
+				if len(data) <= 0 {
+					return errors.New("MissingLocalData")
+				}
+				// XXX fix required base path
```

Comment:
The problematic part is actually the `/stream/<manifestID>` part of the base path. The OS `SaveData` method does this:

```
name, err := os.SaveData("<profile>/<segNo>", data)
```

where `name` is a full(er) path - if using a local OS, it'll look something like `/stream/<manifestID>/<profile>/<seqNo>`

However, the returned `name` is not composable with `SaveData()` itself; it can't be passed back into `SaveData` to over-write existing data. Ideally we could do something like this:

```
name1 = os.SaveData("<profile>/<segNo>", "abc")
name2 = os.SaveData(name1, "def")
name1 == name2
os.GetData(name1) == "def"
```

Ideally the OS itself would do the prefix stripping but that needs to be done on a per-OS basis (since the implementation of `SaveData` is per-OS). Right now we only need it for local OS though. We could also strip the prefix within the `verify` function just prior to saving within the OS, but that's an implementation detail leaking out; the driver API is smelly and could use some deeper thought.

Anyway, worked around the issue within `verify` here https://github.com/livepeer/go-livepeer/pull/1096/commits/7c067159f47ca94b032d8bfd15b2f8997e46d68b